### ISSUE TITLE
Order dependent bug in fine-grained generic alias handling

### ISSUE THREAD

**Issue by msullivan** [python/mypy#4903]
Renaming `b` to `d` in `testAliasFineInitGenericMod` produces the following test, which fails, producing no errors:
```
[case testAliasFineInitGenericMod2]
import c
[file a.py]
from typing import Generic, TypeVar
T = TypeVar('T')
S = TypeVar('S')
class A(Generic[T, S]):
    def __init__(self, x: T) -> None:
        pass
[file a.py.2]
from typing import Generic, TypeVar
T = TypeVar('T')
S = TypeVar('S')
class A(Generic[T, S]):
    def __init__(self, x: S) -> None:
        pass
[file d.py]
import a
B = a.A[int, str]
[file c.py]
from d import B
B(int())
[out]
==
c.py:2: error: Argument 1 has incompatible type "int"; expected "str"
```

**Comment by ilevkivskyi** [comment]
Thanks for finding this!

I actually was afraid this will trigger at some point. The problem is that from the point of view of the snapshot diff, after update `B` points to exactly the same type. A possible solution is to include signatures of `__init__` and `__new__` in alias snapshot, but this may cause some unnecessary re-checks, because aliases are rarely used in runtime context. (On the positive side some existing deps will become redundant if we do this). @JukkaL what do you think? This seems to me similar to the problem of `@abstractmethod` where performance and correctness are also not on the same side.

**Comment by msullivan** [comment]
Would it be possible to create dependencies like:
```
<a.A.__init__> -> <d.B.__init__>
<d.B.__init__> -> x
```?

**Comment by ilevkivskyi** [comment]
I will take a look at this. There might be something more subtle happening.

**Comment by JukkaL** [comment]
This was fixed by #5221.
